crates/ark/src/lsp/main_loop.rs

```diff
@@ -850,27 +932,112 @@ async fn process_diagnostics_batch(batch: Vec<RefreshDiagnosticsTask>) {
 }
 
 pub(crate) fn index_start(folders: Vec<String>, state: WorldState) {
-    INDEXER_QUEUE
-        .send(IndexerQueueTask::Indexer(IndexerTask::Start { folders }))
-        .unwrap_or_else(|err| lsp::log_error!("Failed to queue initial indexing: {err}"));
+    lsp::log_info!("Initial indexing started");
 
+    let uris: Vec<Url> = folders
+        .into_iter()
+        .flat_map(|folder| {
+            walkdir::WalkDir::new(folder)
+                .into_iter()
+                .filter_entry(|e| indexer::filter_entry(e))
+                .filter_map(|entry| {
+                    let entry = match entry {
+                        Ok(e) => e,
+                        Err(_) => return None,
+                    };
+
+                    if !entry.file_type().is_file() {
+                        return None;
+                    }
+                    let path = entry.path();
+
+                    // Only index R files
+                    let ext = path.extension().unwrap_or_default();
+                    if ext != "r" && ext != "R" {
+                        return None;
+                    }
+
+                    if let Ok(uri) = url::Url::from_file_path(path) {
+                        Some(uri)
+                    } else {
+                        tracing::warn!("Can't convert path to URI: {:?}", path);
+                        None
+                    }
+                })
+        })
+        .collect();
+
+    index_create(uris, state);
+}
+
+pub(crate) fn index_create(uris: Vec<Url>, state: WorldState) {
+    for uri in uris {
+        crate::lsp::main_loop::INDEXER_QUEUE
+            .send(crate::lsp::main_loop::IndexerQueueTask::Indexer(
+                crate::lsp::main_loop::IndexerTask::Create { uri },
+            ))
+            .unwrap_or_else(|err| crate::lsp::log_error!("Failed to queue index create: {err}"));
```

Comment:
Use some imports to reduce the number of `::`?

Reply:
It's in scope, no need to qualify! I really try to fix all these agent idiosyncrasies but after a long day I don't see them anymore